src/ast/passes/resource_analyser.cpp

```diff
@@ -307,6 +374,11 @@ void ResourceAnalyser::visit(Map &map)
     resources_.max_read_map_value_size = std::max(
         resources_.max_read_map_value_size, map.type.GetSize());
   }
+  if (needMapKeyAllocation(map) && map_key_exceeds_stack_limit(map)) {
+    resources_.map_key_buffers++;
```

Comment:
Nit: these three lines are repeated. Could be extracted into a helper.

Reply:
Sounds good, I extracted AssignMapStatement and Map parts out to a common function.

lhist/hist/has_key/delete are hard to extract out since they all have slight differences and I found extracting out those parts made it harder to read. Let me know if you disagree though.

Reply:
Might just have to see it first but it's not a big deal either way.